File: solutions/math/172-factorial-trailing-zeroes.py
"""
# 172. Factorial Trailing Zeroes
**Medium**

Given a problem that demonstrates key concepts in Math.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of math concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply math methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages math principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses math techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using math method
3. Return the computed result

</details>
"""

class Solution:
    def trailingZeroes(self, n: int) -> int:
        """
        Count trailing zeroes in n! (factorial).

        Trailing zeroes come from factors of 10 = 2 * 5.
        Since there are always more factors of 2 than 5,
        we just need to count factors of 5.

        Args:
            n: Non-negative integer

        Returns:
            Number of trailing zeroes in n!

        Time Complexity: O(log n)
        Space Complexity: O(1)
        """
        count = 0

        # Count how many numbers <= n are divisible by 5, 25, 125, etc.
        while n >= 5:
            n //= 5
            count += n

        return count

    def solve(self, *args):
        """
        Main solution for 172. Factorial Trailing Zeroes.

        Args:
            *args: Problem-specific arguments

        Returns:
            Number of trailing zeroes

        Time Complexity: O(log n)
        Space Complexity: O(1)
        """
        return self.trailingZeroes(*args)


def test_solution():
    """
    Test cases for 172. Factorial Trailing Zeroes.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    result = solution.trailingZeroes(3)
    expected = 0
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: n = 5 (5! = 120)
    result = solution.trailingZeroes(5)
    expected = 1
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 3: n = 10 (10! has 2 trailing zeroes)
    result = solution.trailingZeroes(10)
    expected = 2
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 4: n = 25 (includes 25 = 5*5)
    result = solution.trailingZeroes(25)
    expected = 6
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 5: n = 0
    result = solution.trailingZeroes(0)
    expected = 0
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 6: Large number
    result = solution.trailingZeroes(100)
    expected = 24
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 7: Power of 5
    result = solution.trailingZeroes(125)
    expected = 31
    assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 172. Factorial Trailing Zeroes")
    for n in [3, 5, 10, 25]:
        print(f"{n}! has {solution.trailingZeroes(n)} trailing zeroes")

File: solutions/math/alternatives/172-factorial-trailing-zeroes.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 *
 * **Step 1:** [description]
 *
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Main solution for Problem 172: Factorial Trailing Zeroes
 *
 * @param {number} n - Non-negative integer
 * @return {number} - Number of trailing zeros in n!
 *
 * Time Complexity: O(log n)
 * Space Complexity: O(1)
 */
function solve(n) {
    let count = 0;

    // Keep dividing n by 5 and add to count
    while (n >= 5) {
        n = Math.floor(n / 5);
        count += n;
    }

    return count;
}

/**
 * Test cases for Problem 172: Factorial Trailing Zeroes
 */
function testSolution() {
    console.log('Testing 172. Factorial Trailing Zeroes');

    // Test case 1: n = 3 (3! = 6, no trailing zeros)
    const result1 = solve(3);
    const expected1 = 0;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: n = 5 (5! = 120, one trailing zero)
    const result2 = solve(5);
    const expected2 = 1;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: n = 10 (10! has 2 trailing zeros)
    const result3 = solve(10);
    const expected3 = 2;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: n = 25 (25! has 6 trailing zeros)
    const result4 = solve(25);
    const expected4 = 6;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    // Test case 5: n = 0
    const result5 = solve(0);
    const expected5 = 0;
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);

    // Test case 6: n = 100
    const result6 = solve(100);
    const expected6 = 24;
    console.assert(result6 === expected6, `Test 6 failed: expected ${expected6}, got ${result6}`);

    // Test case 7: n = 1000
    const result7 = solve(1000);
    const expected7 = 249;
    console.assert(result7 === expected7, `Test 7 failed: expected ${expected7}, got ${result7}`);

    console.log('All test cases passed for 172. Factorial Trailing Zeroes!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 172. Factorial Trailing Zeroes ===');
    console.log('Category: Math');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on math concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
